### PR TITLE
Fix expectations

### PR DESCRIPTION
**Pull request by borismattijssen**
This PR fixes issues with expectations.
It seemed the model of expectations used by k8s was not completely suitable for us.
I implemented a custom version of expectations that works specifically for workflows.

### CODE REVIEW

File: pkg/workflow/utils.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+// StepsToTimestamp is a set of steps holding a timestamp.
+// The timestamp is the time when the expectation for the step was created.
+// This timestamp is used to keep track of a possible expectation timeout.
+type StepsToTimestamp map[string]time.Time
+
+// WorkflowStepExpectations is a way to keep track of expectations for steps of
+// workflows. See "k8s.io/kubernetes/controller/controller_utils.go" for more
+// information on expectations.
```
Comment (nov1n) [inline pkg/workflow/utils.go]: on --> about expectations

File: pkg/workflow/utils.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+type WorkflowStepExpectations struct {
+	// storeLock is used to make use of safe concurrent calls.
+	storeLock sync.Mutex
+	// store maps a workflow key to a set of steps.
```
Comment (nov1n) [inline pkg/workflow/utils.go]: You say it maps to a set of steps but the type is StepsToTimestamp?

File: pkg/workflow/utils.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	// Create step set if it doesn't exist yet.
+	if _, exists := w.store[wfKey]; !exists {
+		w.store[wfKey] = StepsToTimestamp{}
+	}
+	w.store[wfKey][step] = time.Now()
+}
+
+// isExpired returns true if a step expectation has expired.
+func isExpired(timestamp time.Time) bool {
+	return util.RealClock{}.Since(timestamp) > ExpectationsTimeout
```
Comment (nov1n) [inline pkg/workflow/utils.go]: Could use `return timestamp.After(ExpectationsTimeout)`

File: pkg/workflow/utils.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// ExpectationsForStepSatisfied returns whether creation expectations for a
+// step for a workflow are satisfied.
```
Comment (nov1n) [inline pkg/workflow/utils.go]: for a step for a workflow --> for a workflow step

File: pkg/workflow/utils.go (hunk 3 above)
Comment (borismattijssen) [inline pkg/workflow/utils.go]: I don't think so. ExpectationsTimeout is a Duration

**Pull request merged**